### PR TITLE
Support more scheduling policies

### PR DESCRIPTION
**Pull request by Expertium**

### CODE REVIEW

**Comment by Expertium** [comment]
@L-M-Sherlock
Good news: now the simulator supports policies that don't rely on D and S
Bad news: Memrise's incredibly inflexible crap somehow beats scheduling with fixed DR 🤣
So I'd like you to try this code out to see what went wrong

**Comment by Expertium** [comment]
Hm, strange
With the default settings it doesn't outperform fixed DR.
| Memrise | 88.8 | 15.5 | 9680 | 625 |
| DR=0.87 | 76.5 | 13.5 | 9626 | 714 |
| DR=0.88 | 81.3 | 14.1 | 9664 | 685 |
| DR=0.89 | 87.1 | 14.8 | 9700 | 654 |
But if you set learn_span to 5 years and S_max to 25 years, it does.
| Memrise | 97.0 | 18.3 | 9450 | 518 |
| DR=0.87 | 103.0 | 18.8 | 9288 | 495 |
| DR=0.88 | 110.5 | 19.8 | 9349 | 473 |
| DR=0.89 | 118.9 | 20.9 | 9412 | 451 |
| DR=0.90 | 127.9 | 22.0 | 9467 | 430 |

**Comment by Expertium** [comment]
Ok, it's not `s_max_aware_fixed_interval`. Idk how Memrise is better

**Comment by L-M-Sherlock** [comment]
| Scheduling Policy | Average number of reviews per day | Average number of minutes per day | Total knowledge at the end | Knowledge per minute |
| --- | --- | --- | --- | --- |
| SSP-MMC | 60.2 | 11.5 | 9540 | 827 |
| Memrise | 72.2 | 13.1 | 9694 | 739 |
| DR=0.70 | 39.0 | 8.9 | 8605 | 968 |
| DR=0.73 | 43.3 | 9.5 | 8831 | 932 |
| DR=0.76 | 48.2 | 10.1 | 9043 | 895 |
| DR=0.79 | 53.8 | 10.8 | 9228 | 857 |
| DR=0.82 | 60.5 | 11.6 | 9394 | 811 |
| DR=0.85 | 69.1 | 12.6 | 9541 | 757 |
| DR=0.88 | 81.2 | 14.1 | 9665 | 686 |
| DR=0.91 | 100.2 | 16.5 | 9772 | 593 |
| DR=0.94 | 135.7 | 21.1 | 9866 | 468 |
| DR=0.97 | 236.7 | 34.5 | 9943 | 288 |
| IVL=3 | 2199.5 | 298.6 | 9991 | 33 |
| IVL=7 | 959.5 | 132.7 | 9981 | 75 |
| IVL=30 | 198.0 | 31.3 | 9850 | 315 |

**Comment by Expertium** [comment]
Try running it with S_max=25 years and learn_span=5 years

**Comment by L-M-Sherlock** [comment]
I'm working on this: https://github.com/open-spaced-repetition/srs-benchmark/pull/256

So, my device is busy now. Could you give me a full report?

**Comment by Expertium** [comment]
S_MAX = 365 * 25, learn_span=365 * 5
| Memrise | 99.3 | 18.4 | 9461 | 515 |
| DR=0.70 | 48.7 | 11.8 | 8084 | 686 |
| DR=0.71 | 50.4 | 12.0 | 8164 | 679 |
| DR=0.72 | 52.3 | 12.3 | 8243 | 670 |
| DR=0.73 | 54.4 | 12.6 | 8301 | 660 |
| DR=0.74 | 56.5 | 12.9 | 8380 | 651 |
| DR=0.75 | 58.9 | 13.2 | 8484 | 643 |
| DR=0.76 | 61.4 | 13.5 | 8567 | 633 |
| DR=0.77 | 63.9 | 13.8 | 8641 | 624 |
| DR=0.78 | 66.6 | 14.2 | 8728 | 614 |
| DR=0.79 | 69.5 | 14.6 | 8783 | 602 |
| DR=0.80 | 72.4 | 14.9 | 8836 | 591 |
| DR=0.81 | 75.6 | 15.3 | 8907 | 581 |
| DR=0.82 | 79.2 | 15.8 | 8965 | 568 |
| DR=0.83 | 83.0 | 16.3 | 9037 | 555 |
| DR=0.84 | 87.1 | 16.8 | 9109 | 543 |
| DR=0.85 | 92.1 | 17.4 | 9176 | 527 |
| DR=0.86 | 97.5 | 18.1 | 9236 | 510 |
| DR=0.87 | 103.3 | 18.8 | 9288 | 493 |
| DR=0.88 | 110.2 | 19.7 | 9349 | 474 |
| DR=0.89 | 118.0 | 20.7 | 9413 | 454 |
| DR=0.90 | 127.3 | 21.9 | 9466 | 432 |
| DR=0.91 | 138.9 | 23.4 | 9523 | 407 |
| DR=0.92 | 152.6 | 25.2 | 9579 | 380 |
| DR=0.93 | 169.9 | 27.5 | 9634 | 351 |
| DR=0.94 | 192.2 | 30.4 | 9688 | 319 |
| DR=0.95 | 224.0 | 34.6 | 9741 | 281 |
| DR=0.96 | 269.1 | 40.7 | 9793 | 241 |
| DR=0.97 | 343.2 | 50.6 | 9845 | 195 |
| DR=0.98 | 486.5 | 69.9 | 9896 | 142 |
image

**Comment by L-M-Sherlock** [comment]
Why is the Y-axis not the total cost?

**Comment by Expertium** [comment]
I prefer it that way. But sure, here is the _average_ cost (total is not outputted currentlyt, it would require modifying the code):
image

**Comment by L-M-Sherlock** [comment]
How to generate the plots with this PR? It would be nice if you include them in this PR.

**Comment by Expertium** [comment]
Note that my implementation of the hyperparameter optimizer is kinda jank. It calculates the mean and the standard error of the mean (SEM), but that one depends on `PARALLEL`. So if you change `PARALLEL`, you'll get a different SEM, and hence somewhat different results. Once I make the proper procedure for running this on 500 users instead of only on the default parameters, this jank will be removed.

File: script.py [hunk 1]
```diff
@@ -511,71 +536,185 @@ def _get_cost(self, s, d):
...
+    # 1. New cards (prev_interval=0): always start with 1 day
+    # 2. Again grade: reset to 1 day
+    # 3. Hard/Good/Easy: advance in sequence with s_max awareness
+    result = torch.where(
+        is_new_card,  # New cards
+        torch.ones_like(prev_interval),  # Start with 1 day
+        torch.where(
+            grade == 1,  # Again
+            torch.ones_like(prev_interval),  # Reset to 1 day
+            s_max_aware_fixed_interval(stability, difficulty, next_intervals, -w[20])
```
[the remark this answers is not in the document]
Comment (Expertium) [reply script.py:669]: Turns out `PARALLEL` affects the results. I commited a fix

File: README.md [hunk 2]
```diff
@@ -18,46 +27,82 @@ Deck size = 10,000 cards, new cards per day = 10, max. reviews per day = 9,999,
...
+
+"Balanced" roughly corresponds to "Number of memorized cards of DR=83% and efficiency of DR=72%".
+
+Overall, it's clear that SSP-MMC outperforms fixed desired retention, which in turn outperforms fixed intervals. Surprisingly, Memrise's simple algorithm performs remarkably well. We are unsure why.
+
+## Caveats
+
+Currently, the results are based on several assumptions and have several limitations:
+
+1) This is based on the assumption that FSRS is perfectly accurate at predicting the probability of recall
```
Comment (Luc-Mcgrady) [inline README.md:105]: > the results are based on several assumptions

```suggestion
1) FSRS is perfectly accurate at predicting the probability of recall
```

this feels like a tautology.
You made the commit like 2 seconds ago so sorry for pointing it out :joy:

**Comment by Expertium** [comment]
Converge.py now uses a stricter convergence criteria and also tries multiple sets of hyperparameters, to ensure that the matrix converges for all of them. Also, instead of only saving at the end, results of tests are saved in another file after every user.

**Comment by Expertium** [comment]
@L-M-Sherlock PR is ready

**Comment by Expertium** [comment]
@L-M-Sherlock

**Comment by L-M-Sherlock** [comment]
I didn't see any new plots.

**Comment by Expertium** [comment]
I updated readme

File: DR=0.70.png
Comment (L-M-Sherlock) [inline DR=0.70.png:1]: why are these plots not in the `plot/` folder?

Comment (Expertium) [reply DR=0.70.png:1]: welp. I guess I don't know how to use github. Can you fix this for me, please?

Comment (Expertium) [reply DR=0.70.png:1]: Ok, I tried making a folder but couldn't figure out how. Then I tried using Copilot, but it worked on some ancient branch for some reason
So yeah, I would appreciate your help
